### PR TITLE
ISSUE-47 # WireMock - Optional dynamic mock.api.port for downstream s…

### PR DESCRIPTION
**Pull request by authorjapps**
…ervices

### CODE REVIEW

File: src/main/java/org/jsmart/zerocode/core/engine/executor/JsonServiceExecutorImpl.java [hunk 1]
```diff
@@ -32,12 +33,20 @@
...
     @Inject
     private ObjectMapper objectMapper;
+
     @Inject
     SmartUtils smartUtils;
+
     @Inject
     BasicHttpClient httpClient;
+
+    @Inject(optional = true)
```
Comment (nirmalchandra) [inline src/main/java/org/jsmart/zerocode/core/engine/executor/JsonServiceExecutorImpl.java]: +1

File: src/main/java/org/jsmart/zerocode/core/engine/mocker/RestEndPointMocker.java [hunk 2]
```diff
@@ -1,51 +1,119 @@
...
+                requestBuilder.withHeader((String) key, equalTo((String) headersMap.get(key)));
+            }
+        }
+        return requestBuilder;
+    }
+
+    private static ResponseDefinitionBuilder responseBuilder(MockStep mockStep, String jsonBodyRequest) {
+        final ResponseDefinitionBuilder responseBuilder = aResponse()
+                .withStatus(mockStep.getResponse().get("status").asInt())
+                //.withHeader("Content-Type", APPLICATION_JSON) //Can be XMl also for SOAP end points
```
Comment (nirmalchandra) [inline src/main/java/org/jsmart/zerocode/core/engine/mocker/RestEndPointMocker.java]: Can you add response headers too ? This is not super needed, but you can have this as a tech-debt.

File: src/main/java/org/jsmart/zerocode/core/engine/mocker/RestEndPointMocker.java [hunk 3]
```diff
@@ -1,51 +1,119 @@
...
 
+
+    public static void restartWireMock(int dynamicPort) {
+        wireMockServer = new WireMockServer(wireMockConfig().port(dynamicPort)); // <-- Strange
+        /*
+         * Stop the wireMock server if it is running previously due to any other tests.
+         */
+        wireMockServer.stop();
+        wireMockServer.start();
+        WireMock.configureFor("localhost", dynamicPort); // <-- Repetition of PORT was needed, this is a wireMock bug
```
Comment (nirmalchandra) [inline src/main/java/org/jsmart/zerocode/core/engine/mocker/RestEndPointMocker.java]: can you upgrade to `<version>1.58</version>` to see if the port is not repeated?
You can have as a tech-debt.

File: src/main/java/org/jsmart/zerocode/core/engine/mocker/RestEndPointMocker.java [hunk 4]
```diff
@@ -1,51 +1,119 @@
...
+
+        final String bodyJson = mockStep.getBody();
+        // -----------------------------------------------
+        // read request body and set to request builder
+        // -----------------------------------------------
+        if(StringUtils.isNotEmpty(bodyJson)){
+            requestBuilder.withRequestBody(equalToJson(bodyJson));
+        }
+
+        final Map<String, Object> headersMap = mockStep.getHeadersMap();
```
Comment (nirmalchandra) [inline src/main/java/org/jsmart/zerocode/core/engine/mocker/RestEndPointMocker.java]: Can be optimized via injected mapper. But not a big deal as it's the same mapper used.

File: src/test/java/org/jsmart/zerocode/core/engine/mocker/RestEndPointMockerTest.java [hunk 5]
```diff
@@ -119,6 +135,84 @@ public void willMockASimpleGetEndPoint() throws Exception{
...
+        request.addHeader("key", "key-007");
+        request.addHeader("secret", "secret-007");
+
+        HttpResponse response = httpClient.execute(request);
+        final String responseBodyActual = IOUtils.toString(response.getEntity().getContent(), "UTF-8");
+        System.out.println("### response: \n" + responseBodyActual);
+
+        assertThat(response.getStatusLine().getStatusCode(), is(200));
+        JSONAssert.assertEquals(jsonBodyResponse, responseBodyActual, true);
+
```
Comment (nirmalchandra) [inline src/test/java/org/jsmart/zerocode/core/engine/mocker/RestEndPointMockerTest.java]: can you cover an error scenario where one of the mocks didnt mock as expected?

File: src/main/java/org/jsmart/zerocode/core/engine/mocker/RestEndPointMocker.java (hunk 4 above)
Comment (authorjapps) [reply src/main/java/org/jsmart/zerocode/core/engine/mocker/RestEndPointMocker.java]: ok. will take care.

File: src/test/java/org/jsmart/zerocode/core/engine/mocker/RestEndPointMockerTest.java (hunk 5 above)
Comment (authorjapps) [reply src/test/java/org/jsmart/zerocode/core/engine/mocker/RestEndPointMockerTest.java]: done. See `org.jsmart.zerocode.wiremock.WireMockIntegrationTest`

File: src/main/java/org/jsmart/zerocode/core/engine/mocker/RestEndPointMocker.java (hunk 3 above)
Comment (authorjapps) [reply src/main/java/org/jsmart/zerocode/core/engine/mocker/RestEndPointMocker.java]: #49 has been raised to deal with this.

**Pull request merged**